devito/finite_differences/differentiable.py

```diff
@@ -803,9 +815,13 @@ def _evaluate(self, **kwargs):
         return EvalDerivative(expr, base=self.base)
 
 
+class DiffDerivative(IndexDerivative, DifferentiableOp):
```

Comment:
I tried all sort of ways; remember we discussed a sympy issue a few days ago, and eventually I landed there...

So, `IndexDerivative` was (master branch) a `DifferentiableOp`, since `IndexSum` was (master branch) a `DifferentiableOp`. The heart of the problem is: if an object is non differentiable but one of its arguments is, then a reconstruction will make the parent object differentiable as well

This new structuring distinguishes between DiffDerivative (differentiable) and the lower-level IndexDerivative (non-differentiable). This ensures that, once inside the compiler, everything will have been lowered such that no Differentiable objects can ever be (re)constructed, which in essence will make compilation behave as expected, having ruled out by construction any funky-ness due to the way SymPy handles subclasses and reconstruction

Does it make sense or does it sound like a ranting? I can elaborate or explain in a call if necessary